openedx/features/enterprise_support/api.py

```diff
@@ -430,34 +468,56 @@ def enterprise_customer_uuid_for_request(request):
         except EnterpriseCustomer.DoesNotExist:
             enterprise_customer_uuid = None
     else:
-        # Check if we got an Enterprise UUID passed directly as either a query
-        # parameter, or as a value in the Enterprise cookie.
-        enterprise_customer_uuid = request.GET.get('enterprise_customer') or request.COOKIES.get(
-            settings.ENTERPRISE_CUSTOMER_COOKIE_NAME
-        )
+        enterprise_customer_uuid = _customer_uuid_from_query_param_cookies_or_session(request)
 
-    if not enterprise_customer_uuid and request.user.is_authenticated:
+    if enterprise_customer_uuid is _CACHE_MISS and request.user.is_authenticated:
         # If there's no way to get an Enterprise UUID for the request, check to see
         # if there's already an Enterprise attached to the requesting user on the backend.
         learner_data = get_enterprise_learner_data_from_db(request.user)
         if learner_data:
-            enterprise_customer_uuid = learner_data[0]['enterprise_customer']['uuid']
+            enterprise_customer = learner_data[0]['enterprise_customer']
+            enterprise_customer_uuid = enterprise_customer['uuid']
+            cache_enterprise(enterprise_customer)
+        else:
+            enterprise_customer_uuid = None
+
+        # Now that we've asked the database for this users's enterprise customer data,
+        # add it to their session (even if it's null/empty, which indicates the user
+        # has no associated enterprise customer).
+        add_enterprise_customer_to_session(request, enterprise_customer)
 
     return enterprise_customer_uuid
 
 
+def _customer_uuid_from_query_param_cookies_or_session(request):
+    """
+    Helper function that plucks a customer UUID out of the given requests's
+    query params, cookie, or session data.
+    Returns ``__CACHE_MISS__`` if none of those keys are present in the request.
+    """
+    for function in (
+        enterprise_customer_uuid_from_query_param,
+        enterprise_customer_uuid_from_cookie,
+        enterprise_customer_uuid_from_session,
```

Comment:
Is there a preference order on these? For example if the session doesn't match the query param, should we be preferring the query param like this currently has it?

Reply:
I kept the order that was previously used in `enterprise_customer_uuid_for_request()` - query param then cookie, then added the session in last, since it's a little more "special" and can only be set from Django.  I think that's the correct order:
1. query param is given higher precedence, since that's what the client is asking for.
2. then check the cookie, because it's used by the front-end code, too.
3. then session, since only Django sets that.

If the param and the session or cookie don't match, I don't know what we would do - you're certainly allowed to ask about other UUIDs, even if you don't have permissions for them.  I think the bad case is if the cookie and the session don't match.  But I'm not sure what to do in that case, or if we care that much about resolving it.